File: src/commands/apply.test.ts
import { afterEach, beforeEach, describe, it } from "@std/testing/bdd";
import { assertSpyCall } from "@std/testing/mock";
import { assertEquals, assertInstanceOf } from "@std/assert";
import { copy, ensureDir } from "@std/fs";
import apply from "./apply.ts";
import {
  assertArrayIncludesRemoveMatch,
  assertSpyLoggerCalls,
  makeSpyLogger,
  type SpyLogger,
} from "../core/test-utils.ts";
import { dirname } from "@std/path";

let workingDir: string;
let logger: SpyLogger;

beforeEach(async () => {
  workingDir = await Deno.makeTempDir();
  logger = makeSpyLogger();
});

afterEach(async () => {
  await Deno.remove(workingDir, { recursive: true });
});

describe("apply()", () => {
  it("should normalize `from` path", async () => {
    const result = await apply({
      from: `${workingDir}/foo/bar/..`,
      extra: { homedir: workingDir, dotfilesFilesDir: workingDir },
    });

    assertEquals(
      (result as Error).message,
      `Path ${`${workingDir}/foo`} does not exists or is not readable`,
    );
  });

  it("should return an error if `from` does not exists", async () => {
    const result = await apply({
      from: `${workingDir}/foo/`,
      extra: { homedir: workingDir, dotfilesFilesDir: workingDir },
    });

    assertInstanceOf(result, Error);
    assertEquals(
      result.message,
      `Path ${`${workingDir}/foo`} does not exists or is not readable`,
    );
  });

  it("should return an error if `from` is not readable", async () => {
    const from = await Deno.makeTempDir({ dir: workingDir });

    await Deno.chmod(from, 0o333);

    await using _ = {
      [Symbol.asyncDispose]: async () => {
        await Deno.chmod(from, 0o755);
      },
    };

    const result = await apply({
      from,
      extra: { homedir: workingDir, dotfilesFilesDir: workingDir },
    });

    assertInstanceOf(result, Error);
    assertEquals(
      result.message,
      `Path ${from} does not exists or is not readable`,
    );
  });

  it("should return an error if `from` is not a subdirectory of `dotfilesFilesDir`", async () => {
    const homedir = await Deno.makeTempDir({ dir: workingDir });
    const dotfilesFilesDir = await Deno.makeTempDir({ dir: workingDir });
    const from = await Deno.makeTempDir({ dir: workingDir });

    const result = await apply({
      from,
      extra: { homedir, dotfilesFilesDir },
    });

    assertInstanceOf(result, Error);
    assertEquals(
      result.message,
      `Path ${from} is not a subpath of ${dotfilesFilesDir}`,
    );
  });

  it("should apply a single file", async () => {
    const homedir = await Deno.makeTempDir({ dir: workingDir });
    const dotfilesFilesDir = await Deno.makeTempDir({ dir: workingDir });
    const from = await Deno.makeTempFile({ dir: dotfilesFilesDir });

    const result = await apply({
      from,
      extra: { homedir, dotfilesFilesDir, logger },
    });

    assertEquals(result, true);
    assertSpyLoggerCalls(logger, { logNoNl: { log: 1 }, log: { log: 1 } });
    assertSpyCall(logger.logNoNl.log, 0, {
      args: [
        `Applying ${from} to ${from.replace(dotfilesFilesDir, homedir)} ...`,
      ],
    });
    assertSpyCall(logger.log.log, 0, { args: [" ✓"] });
  });

  it("should return an error if something appens while applying file", async () => {
    const homedir = await Deno.makeTempDir({ dir: workingDir });
    const dotfilesFilesDir = await Deno.makeTempDir({ dir: workingDir });
    const from = await Deno.makeTempFile({ dir: dotfilesFilesDir });
    const to = from.replace(dotfilesFilesDir, homedir);

    await using _ = await Deno.create(to);
    await Deno.chmod(to, 0o444);

    const result = await apply({
      from,
      extra: { homedir, dotfilesFilesDir, logger },
    });

    assertInstanceOf(result, Error);
    assertEquals(result.message, `Error applying ${from} to ${to}`);
    assertInstanceOf(result.cause, Error);
    assertSpyLoggerCalls(logger, { logNoNl: { log: 1 }, log: { log: 1 } });
    assertSpyCall(logger.logNoNl.log, 0, {
      args: [
        `Applying ${from} to ${from.replace(dotfilesFilesDir, homedir)} ...`,
      ],
    });
    assertSpyCall(logger.log.log, 0, { args: [" ✕"] });
  });

  it("should fallback to `dotfilesFilesDir` if `from` is not provided", async () => {
    const homedir = await Deno.makeTempDir({ dir: workingDir });
    const dotfilesFilesDir = await Deno.makeTempDir({ dir: workingDir });

    await Deno.chmod(dotfilesFilesDir, 0o333);

    await using _ = {
      [Symbol.asyncDispose]: async () => {
        await Deno.chmod(dotfilesFilesDir, 0o755);
      },
    };

    const result = await apply({
      extra: { homedir, dotfilesFilesDir },
    });

    assertInstanceOf(result, Error);
    assertEquals(
      result.message,
      `Path ${dotfilesFilesDir} does not exists or is not readable`,
    );
  });

  it("should apply a directory without providing `from`", async () => {
    const homedir = await Deno.makeTempDir({ dir: workingDir });
    const dotfilesFilesDir = await Deno.makeTempDir({ dir: workingDir });
    const f1 = await Deno.makeTempFile({ dir: dotfilesFilesDir, prefix: "1-" });
    const f2 = await Deno.makeTempFile({ dir: dotfilesFilesDir, prefix: "2-" });

    const result = await apply({
      extra: { homedir, dotfilesFilesDir, logger },
    });

    assertEquals(result, true);
    assertSpyLoggerCalls(logger, { logNoNl: { log: 2 }, log: { log: 2 } });
    assertArrayIncludesRemoveMatch([
      ...logger.logNoNl.log.calls[0].args,
      ...logger.logNoNl.log.calls[1].args,
    ], [
      `Applying ${f1} to ${f1.replace(dotfilesFilesDir, homedir)} ...`,
      `Applying ${f2} to ${f2.replace(dotfilesFilesDir, homedir)} ...`,
    ]);
    assertArrayIncludesRemoveMatch([
      ...logger.log.log.calls[0].args,
      ...logger.log.log.calls[1].args,
    ], [
      " ✓",
      " ✓",
    ]);
  });

  it("should apply a directory", async () => {
    const homedir = await Deno.makeTempDir({ dir: workingDir });
    const dotfilesFilesDir = await Deno.makeTempDir({ dir: workingDir });
    const from = await Deno.makeTempDir({ dir: dotfilesFilesDir });
    const f1 = await Deno.makeTempFile({ dir: from, prefix: "1-" });
    const f2 = await Deno.makeTempFile({ dir: from, prefix: "2-" });

    const result = await apply({
      from,
      extra: { homedir, dotfilesFilesDir, logger },
    });

    assertEquals(result, true);
    assertSpyLoggerCalls(logger, { logNoNl: { log: 2 }, log: { log: 2 } });
    assertArrayIncludesRemoveMatch([
      ...logger.logNoNl.log.calls[0].args,
      ...logger.logNoNl.log.calls[1].args,
    ], [
      `Applying ${f1} to ${f1.replace(dotfilesFilesDir, homedir)} ...`,
      `Applying ${f2} to ${f2.replace(dotfilesFilesDir, homedir)} ...`,
    ]);
    assertArrayIncludesRemoveMatch([
      ...logger.log.log.calls[0].args,
      ...logger.log.log.calls[1].args,
    ], [
      " ✓",
      " ✓",
    ]);
  });

  it("should return an error if something appens while applying directory", async () => {
    const homedir = await Deno.makeTempDir({ dir: workingDir });
    const dotfilesFilesDir = await Deno.makeTempDir({ dir: workingDir });
    const from = await Deno.makeTempDir({ dir: dotfilesFilesDir });
    const f1 = await Deno.makeTempFile({ dir: from, prefix: "1-" });
    const f2 = await Deno.makeTempFile({ dir: from, prefix: "2-" });

    await ensureDir(dirname(f2.replace(dotfilesFilesDir, homedir)));
    await copy(f1, f2.replace(dotfilesFilesDir, homedir));
    await Deno.chmod(f2.replace(dotfilesFilesDir, homedir), 0o444);

    const result = await apply({
      from,
      extra: { homedir, dotfilesFilesDir, logger },
    });

    assertInstanceOf(result, AggregateError);
    assertEquals(result.message, "Error applying directory");
    assertEquals(result.errors.length, 1);
    assertInstanceOf(result.errors[0], Error);
    assertEquals(
      result.errors[0].message,
      `Error applying ${f2} to ${f2.replace(dotfilesFilesDir, homedir)}`,
    );
    assertInstanceOf(result.errors[0].cause, Error);
    assertSpyLoggerCalls(logger, { logNoNl: { log: 2 }, log: { log: 2 } });
    assertArrayIncludesRemoveMatch([
      ...logger.logNoNl.log.calls[0].args,
      ...logger.logNoNl.log.calls[1].args,
    ], [
      `Applying ${f1} to ${f1.replace(dotfilesFilesDir, homedir)} ...`,
      `Applying ${f2} to ${f2.replace(dotfilesFilesDir, homedir)} ...`,
    ]);
    assertArrayIncludesRemoveMatch([
      ...logger.log.log.calls[0].args,
      ...logger.log.log.calls[1].args,
    ], [
      " ✓",
      " ✕",
    ]);
  });
});

File: src/commands/apply_test.go
package commands_test

import (
	"fmt"
	"os"
	"strings"

	"github.com/fatih/color"
	"github.com/m4rc3l05/dots/src/commands"
	"github.com/m4rc3l05/dots/src/core"
	"github.com/m4rc3l05/dots/src/testing"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

var _ = Describe("adopt()", func() {
	var workingDir string
	var logger *testing.SpyLogger
	var cmd commands.Commands

	_ = BeforeEach(func() {
		workingDir, _ = os.MkdirTemp(os.TempDir(), "wd-*")
		logger = testing.MakeSpyLogger()
		color.NoColor = true
		cmd = commands.Commands{Logger: logger}
	})

	_ = AfterEach(func() {
		os.RemoveAll(workingDir)
	})

	It("should normalize `from` path", func() {
		result, err := cmd.Apply(commands.ApplyArgs{
			From: "/foo/bar/..",
			Extra: commands.ApplyArgsExtra{
				Homedir:          workingDir,
				DotfilesFilesDir: workingDir,
			},
		})

		Expect(result).To(BeFalse())
		Expect(
			err,
		).To(MatchError(fmt.Sprintf("path %s does not exists or is not readable", "/foo")))
		testing.AssertSpyLoggerCalls(*logger, nil)
	})

	It("should return an error if `from` does not exists", func() {
		result, err := cmd.Apply(commands.ApplyArgs{
			From: workingDir + "/foo",
			Extra: commands.ApplyArgsExtra{
				Homedir:          workingDir,
				DotfilesFilesDir: workingDir,
			},
		})

		Expect(result).To(BeFalse())
		Expect(
			err,
		).To(MatchError(fmt.Sprintf("path %s does not exists or is not readable", workingDir+"/foo")))
		testing.AssertSpyLoggerCalls(*logger, nil)
	})

	It("should return an error if `from` is not readable", func() {
		from, _ := os.MkdirTemp(workingDir, "*")

		os.Chmod(from, 0o300)

		result, err := cmd.Apply(commands.ApplyArgs{
			From: from,
			Extra: commands.ApplyArgsExtra{
				Homedir:          workingDir,
				DotfilesFilesDir: workingDir,
			},
		})

		Expect(result).To(BeFalse())
		Expect(err).To(MatchError(fmt.Sprintf("path %s does not exists or is not readable", from)))
		testing.AssertSpyLoggerCalls(*logger, nil)
	})

	It("should return an error if `from` is not a subdirectory of `dotfilesFilesDir`", func() {
		homedir, _ := os.MkdirTemp(workingDir, "*")
		dotfilesFilesDir, _ := os.MkdirTemp(workingDir, "*")
		from, _ := os.MkdirTemp(workingDir, "*")

		result, err := cmd.Apply(commands.ApplyArgs{
			From: from,
			Extra: commands.ApplyArgsExtra{
				Homedir:          homedir,
				DotfilesFilesDir: dotfilesFilesDir,
			},
		})

		Expect(result).To(BeFalse())
		Expect(
			err,
		).To(MatchError(fmt.Sprintf("path %s is not a subpath of %s", from, dotfilesFilesDir)))
		testing.AssertSpyLoggerCalls(*logger, nil)
	})

	It("should apply a single file", func() {
		homedir, _ := os.MkdirTemp(workingDir, "*")
		dotfilesFilesDir, _ := os.MkdirTemp(workingDir, "*")
		from, _ := os.CreateTemp(dotfilesFilesDir, "*")
		defer from.Close()

		result, err := cmd.Apply(commands.ApplyArgs{
			From: from.Name(),
			Extra: commands.ApplyArgsExtra{
				Homedir:          homedir,
				DotfilesFilesDir: dotfilesFilesDir,
			},
		})

		Expect(result).To(BeTrue())
		Expect(err).To(BeNil())
		testing.AssertSpyLoggerCalls(*logger, &testing.SpyLoggerCallNumber{Log: 1, Lognl: 1})
		Expect(
			logger.Calls.Log[0].Args,
		).To(Equal([]any{"Applying %s to %s ...", from.Name(), strings.Replace(from.Name(), dotfilesFilesDir, homedir, 1)}))
		Expect(logger.Calls.Lognl[0].Args).To(Equal([]any{" ✓"}))
	})

	It("should return an error if something appens while applying file", func() {
		homedir, _ := os.MkdirTemp(workingDir, "*")
		dotfilesFilesDir, _ := os.MkdirTemp(workingDir, "*")
		from, _ := os.CreateTemp(dotfilesFilesDir, "*")
		defer from.Close()
		to, _ := os.Create(strings.Replace(from.Name(), dotfilesFilesDir, homedir, 1))
		defer to.Close()

		os.Chmod(to.Name(), 0o400)

		result, err := cmd.Apply(commands.ApplyArgs{
			From: from.Name(),
			Extra: commands.ApplyArgsExtra{
				Homedir:          homedir,
				DotfilesFilesDir: dotfilesFilesDir,
			},
		})

		Expect(result).To(BeFalse())
		unwrapErrors, _ := err.(interface{ Unwrap() []error })
		Expect(
			unwrapErrors.Unwrap()[0],
		).To(MatchError(fmt.Sprintf("error applying %s to %s", from.Name(), to.Name())))
		Expect(
			unwrapErrors.Unwrap()[1],
		).To(MatchError(fmt.Sprintf("open %s: permission denied", to.Name())))
		testing.AssertSpyLoggerCalls(*logger, &testing.SpyLoggerCallNumber{Log: 1, Lognl: 1})
		Expect(
			logger.Calls.Log[0].Args,
		).To(Equal([]any{"Applying %s to %s ...", from.Name(), strings.Replace(from.Name(), dotfilesFilesDir, homedir, 1)}))
		Expect(logger.Calls.Lognl[0].Args).To(Equal([]any{" ✕"}))
	})

	It("should apply a directory", func() {
		homedir, _ := os.MkdirTemp(workingDir, "*")
		dotfilesFilesDir, _ := os.MkdirTemp(workingDir, "*")
		f1, _ := os.CreateTemp(dotfilesFilesDir, "1-*")
		defer f1.Close()
		f2, _ := os.CreateTemp(dotfilesFilesDir, "2-*")
		defer f2.Close()

		result, err := cmd.Apply(commands.ApplyArgs{
			From: dotfilesFilesDir,
			Extra: commands.ApplyArgsExtra{
				Homedir:          homedir,
				DotfilesFilesDir: dotfilesFilesDir,
			},
		})

		Expect(result).To(BeTrue())
		Expect(err).To(BeNil())
		testing.AssertSpyLoggerCalls(*logger, &testing.SpyLoggerCallNumber{Log: 2, Lognl: 2})
		Expect(
			logger.Calls.Log[0].Args,
		).To(Equal([]any{"Applying %s to %s ...", f1.Name(), strings.Replace(f1.Name(), dotfilesFilesDir, homedir, 1)}))
		Expect(
			logger.Calls.Log[1].Args,
		).To(Equal([]any{"Applying %s to %s ...", f2.Name(), strings.Replace(f2.Name(), dotfilesFilesDir, homedir, 1)}))
		Expect(logger.Calls.Lognl[0].Args).To(Equal([]any{" ✓"}))
		Expect(logger.Calls.Lognl[1].Args).To(Equal([]any{" ✓"}))
	})

	It("should return an error if something appens while applying directory", func() {
		homedir, _ := os.MkdirTemp(workingDir, "*")
		dotfilesFilesDir, _ := os.MkdirTemp(workingDir, "*")
		f1, _ := os.CreateTemp(dotfilesFilesDir, "1-*")
		defer f1.Close()
		f2, _ := os.CreateTemp(dotfilesFilesDir, "2-*")
		defer f2.Close()

		core.RecreateFile(f2.Name(), strings.Replace(f2.Name(), dotfilesFilesDir, homedir, 1))
		os.Chmod(strings.Replace(f2.Name(), dotfilesFilesDir, homedir, 1), 0o400)

		result, err := cmd.Apply(commands.ApplyArgs{
			From: dotfilesFilesDir,
			Extra: commands.ApplyArgsExtra{
				Homedir:          homedir,
				DotfilesFilesDir: dotfilesFilesDir,
			},
		})

		Expect(result).To(BeFalse())
		unwrapErrors, _ := err.(interface{ Unwrap() []error })
		Expect(unwrapErrors.Unwrap()[0]).To(MatchError("error applying directory"))
		unwrapErrors, _ = unwrapErrors.Unwrap()[1].(interface{ Unwrap() []error })
		Expect(
			unwrapErrors.Unwrap()[0],
		).To(MatchError(fmt.Sprintf("error applying %s to %s", f2.Name(), strings.Replace(f2.Name(), dotfilesFilesDir, homedir, 1))))
		Expect(
			unwrapErrors.Unwrap()[1],
		).To(MatchError(fmt.Sprintf("open %s: permission denied", strings.Replace(f2.Name(), dotfilesFilesDir, homedir, 1))))
		testing.AssertSpyLoggerCalls(*logger, &testing.SpyLoggerCallNumber{Log: 2, Lognl: 2})
		Expect(
			logger.Calls.Log[0].Args,
		).To(Equal([]any{"Applying %s to %s ...", f1.Name(), strings.Replace(f1.Name(), dotfilesFilesDir, homedir, 1)}))
		Expect(
			logger.Calls.Log[1].Args,
		).To(Equal([]any{"Applying %s to %s ...", f2.Name(), strings.Replace(f2.Name(), dotfilesFilesDir, homedir, 1)}))
		Expect(logger.Calls.Lognl[0].Args).To(Equal([]any{" ✓"}))
		Expect(logger.Calls.Lognl[1].Args).To(Equal([]any{" ✕"}))
	})
})
